src/ZArray.jl

```diff
@@ -138,81 +145,145 @@ maybeinner(a::Array) = a
+    write_items!(z.storage,writechannel,z.path,blockr)
+  end
+  bind(writechannel,writetask)
+  
+  try 
+    for i in 1:length(blockr)
+      
+      bI,chunk_compressed = take!(readchannel)
+      
+      current_chunk_offsets = map((s,i)->s*(i-1),size(a),Tuple(bI))
+
+      indranges    = map(boundint,r.indices,size(a),current_chunk_offsets)
+
+      if isnothing(chunk_compressed) || (length.(indranges) != size(a))
+        resetbuffer!(z.metadata.fill_value,a)
+      end
+
+      curchunk = if length.(indranges) != size(a)
+        view(a,dotminus.(indranges,current_chunk_offsets)...)
+      else
+        a
+      end
+      
+      if chunk_compressed !== nothing
+        uncompress_raw!(a,z,chunk_compressed)
+      end
+
+      curchunk .= view(ain,dotminus.(indranges,input_base_offsets)...)
+
+      put!(writechannel,bI=>compress_raw(maybeinner(a),z))
+      nothing
     end
+  finally
+    close(readchannel)
+    close(writechannel)
   end
-  aout
+  ain
 end
 
 DiskArrays.readblock!(a::ZArray,aout,i::AbstractUnitRange...) = readblock!(aout,a,CartesianIndices(i))
-DiskArrays.writeblock!(a::ZArray,v,i::AbstractUnitRange...) = readblock!(v,a,CartesianIndices(i),readmode=false)
+DiskArrays.writeblock!(a::ZArray,v,i::AbstractUnitRange...) = writeblock!(v,a,CartesianIndices(i))
 DiskArrays.haschunks(::ZArray) = DiskArrays.Chunked()
 DiskArrays.eachchunk(a::ZArray) = DiskArrays.GridChunks(a,a.metadata.chunks)
 
 """
-    readchunk!(a::DenseArray{T},z::ZArray{T,N},i::CartesianIndex{N})
+uncompress_raw!(a::DenseArray{T},z::ZArray{T,N},i::CartesianIndex{N})
 
 Read the chunk specified by `i` from the Zarray `z` and write its content to `a`
 """
-function readchunk!(a::DenseArray,z::ZArray{<:Any,N},i::CartesianIndex{N}) where N
-    length(a) == prod(z.metadata.chunks) || throw(DimensionMismatch("Array size does not equal chunk size"))
-    curchunk = z.storage[z.path,i]
-    if curchunk === nothing
-        fill!(a, z.metadata.fill_value)
-    else
-        zuncompress!(a, curchunk, z.metadata.compressor, z.metadata.filters)
-    end
-    a
+function uncompress_raw!(a,z::ZArray{<:Any,N},curchunk) where N
+  if curchunk === nothing
+    fill!(a, z.metadata.fill_value)
+  else
+    zuncompress!(a, curchunk, z.metadata.compressor, z.metadata.filters)
+  end
+  a
 end
 
-"""
-    writechunk!(a::DenseArray{T},z::ZArray{T,N},i::CartesianIndex{N}) where {T,N}
+dotminus(x,y) = x.-y
 
-Write the data from the array `a` to the chunk `i` in the ZArray `z`
-"""
-function writechunk!(a::DenseArray, z::ZArray{<:Any,N}, i::CartesianIndex{N}) where N
-  z.writeable || error("Can not write to read-only ZArray")
-  z.writeable || error("ZArray not in write mode")
+function uncompress_to_output!(aout,output_base_offsets,z,chunk_compressed,current_chunk_offsets,a,indranges)
+  curchunk = view(a,dotminus.(indranges,current_chunk_offsets)...)
+  uncompress_raw!(a,z,chunk_compressed)
+  aout[dotminus.(indranges, output_base_offsets)...] = curchunk
```

Comment:
This portion of the code remains a major bottleneck, the copying can still be made much more efficient with the following suggested change (nearly 3x faster in my tests)

```suggestion
  uncompress_raw!(a,z,chunk_compressed)
  aout[dotminus.(indranges, output_base_offsets)...] = a[dotminus.(indranges,current_chunk_offsets)...]
```

Reply:
Thanks a lot. It really looks like this is one of the cases where allocating an extra array improves performance. The main issue I have here is that we need the view/copy into the buffer array only for incomplete chunks, so in most cases `aout[dotminus.(indranges, output_base_offsets)...] = a` is sufficient and most performant. So maybe special-casing for full/partial chunk and copying for partial chunks should be best.

Reply:
Great point. I have updated the suggestion to conditionally do a partial copy only for incomplete chunks.